src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs

```diff
@@ -66,23 +72,67 @@ internal HttpInMetricsListener(string name, Meter meter, AspNetCoreMetricsInstru
     {
-        if (name == OnStopEvent)
+        switch (name)
         {
-            if (this.emitOldAttributes)
-            {
-                this.OnEventWritten_Old(name, payload);
-            }
+            case OnUnhandledDiagnosticsExceptionEvent:
+            case OnUnhandledHostingExceptionEvent:
+                {
+                    if (this.emitNewAttributes)
+                    {
+                        this.OnExceptionEventWritten(name, payload);
+                    }
+                }
+
+                break;
+            case OnStopEvent:
+                {
+                    if (this.emitOldAttributes)
+                    {
+                        this.OnEventWritten_Old(name, payload);
+                    }
+
+                    if (this.emitNewAttributes)
+                    {
+                        this.OnEventWritten_New(name, payload);
+                    }
+                }
+
+                break;
+        }
+    }
 
-            if (this.emitNewAttributes)
-            {
-                this.OnEventWritten_New(name, payload);
-            }
+    public void OnExceptionEventWritten(string name, object payload)
+    {
+        // We need to use reflection here as the payload type is not a defined public type.
+        if (!TryFetchException(payload, out Exception exc) || !TryFetchHttpContext(payload, out HttpContext ctx))
+        {
+            AspNetCoreInstrumentationEventSource.Log.NullPayload(nameof(HttpInListener), nameof(this.OnExceptionEventWritten), HttpServerDurationMetricName);
+            return;
         }
+
+        ctx.Items.Add(ErrorTypekey, exc.GetType().FullName);
+
+        // See https://github.com/dotnet/aspnetcore/blob/690d78279e940d267669f825aa6627b0d731f64c/src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs#L252
+        // and https://github.com/dotnet/aspnetcore/blob/690d78279e940d267669f825aa6627b0d731f64c/src/Middleware/Diagnostics/src/DeveloperExceptionPage/DeveloperExceptionPageMiddlewareImpl.cs#L174
+        // this makes sure that top-level properties on the payload object are always preserved.
+#if NET6_0_OR_GREATER
+        [UnconditionalSuppressMessage("Trimming", "IL2026", Justification = "The event source guarantees that top level properties are preserved")]
+#endif
+        static bool TryFetchException(object payload, out Exception exc)
+            => ExceptionPropertyFetcher.TryFetch(payload, out exc) && exc != null;
+
+#if NET6_0_OR_GREATER
+        [UnconditionalSuppressMessage("Trimming", "IL2026", Justification = "The event source guarantees that top level properties are preserved")]
+#endif
+        static bool TryFetchHttpContext(object payload, out HttpContext ctx)
+            => HttpContextPropertyFetcher.TryFetch(payload, out ctx) && ctx != null;
     }
 
     public void OnEventWritten_Old(string name, object payload)
     {
         var context = payload as HttpContext;
+        context.Items.TryGetValue(name, out var item);
```

Comment:
what is this? i don't see `item` being used

Reply:
good catch. this was unintended.